controllers/phases.go

```diff
@@ -0,0 +1,477 @@
+	}
+
+	// Generate a set of new objects using the clusterctl library. NewComponents() will do the yaml proccessing,
+	// like ensure all the provider components are in proper namespace, replcae variables, etc. See the clusterctl
+	// documentation for more details.
+	p.components, err = repository.NewComponents(repository.ComponentsInput{
+		Provider:     p.providerConfig,
+		ConfigClient: p.configClient,
+		Processor:    yamlprocessor.NewSimpleProcessor(),
+		RawYaml:      componentsFile,
+		Options:      p.options})
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, operatorv1.ComponentsFetchErrorReason, operatorv1.PreflightCheckCondition)
+	}
+
+	// ProviderSpec provides fields for customizing the provider deployment options.
+	// We can use clusterctl library to apply this customizations.
+	err = repository.AlterComponents(p.components, customizeObjectsFn(p.provider))
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, operatorv1.ComponentsFetchErrorReason, operatorv1.PreflightCheckCondition)
+	}
+
+	conditions.Set(p.provider, conditions.TrueCondition(operatorv1.PreflightCheckCondition))
+	return reconcile.Result{}, nil
+}
+
+// preInstall ensure all the clusterctl CRDs are available before installing the provider,
+// and delete existing components if required for upgrade.
+func (p *phaseReconciler) preInstall(ctx context.Context) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	clusterClient := p.newClusterClient()
+
+	log.V(2).Info("Ensuring clustectl CRDs are installed", "name", p.provider.GetName())
+	err := clusterClient.ProviderInventory().EnsureCustomResourceDefinitions()
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, "failed installing clusterctl CRDs", operatorv1.ProviderInstalledCondition)
+	}
+
+	needPreDelete, err := p.updateRequiresPreDeletion(ctx, p.provider)
+	if err != nil || !needPreDelete {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, "failed getting clusterctl Provider", operatorv1.ProviderInstalledCondition)
+	}
+	return p.delete(ctx)
+}
+
+// updateRequiresPreDeletion get the clusterctl Provider and compare it's version with the Spec.Version
+// if different, it's an upgrade.
+func (s *phaseReconciler) updateRequiresPreDeletion(ctx context.Context, provider genericprovider.GenericProvider) (bool, error) {
+	// TODO: We should replace this with an Installed/Applied version in the providerStatus.
+	err := s.ctrlClient.Get(ctx, clusterctlProviderName(provider), s.clusterctlProvider)
+	if apierrors.IsNotFound(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	if s.clusterctlProvider.Version == "" {
+		return false, nil
+	}
+	nextVersion := versionutil.MustParseSemantic(s.components.Version())
+	return versionutil.MustParseSemantic(s.clusterctlProvider.Version).LessThan(nextVersion), nil
```

Comment:
Why `MustParse`, that suggests to me that it will panic if there's an issue? Why not return an error?

Reply:
The error will break the chain pattern, I'd like to use this function, it looks like there is no other way.

Reply:
This function is already returning an error though? So it would be a relatively straight forward change
```
nextVersion, err := versionutil.ParseSemantic(s.components.Version())
if err != nil {
  return false, err
}
currentVersion, err := versionutil.MustParse(s.clusterctlProvider.Version)
if err != nil {
  return false, err
}
return currentVersion.LessThan(nextVersion), nil
```